### PR TITLE
Add output type

### PR DESCRIPTION
**Pull request by mollyxu**
## Summary

First of three PRs adding lossless HDR decoding support. This PR adds only the plumbing — no C++ output behavior changes. The default stays `uint8`, so all existing tests pass unchanged.

 - New `OutputDtype` enum (`UINT8` / `FLOAT32` / `AUTO`) in `StreamOptions.h`, plus a matching field on `VideoStreamOptions` (default `UINT8`, so existing callers are unchanged).        
  - `_add_video_stream` / `add_video_stream` schemas in `custom_ops.cpp` gain an optional `output_dtype: str?` arg. Parses `"uint8"` / `"float32"` / `"auto"` into the enum and raises     
  `"Invalid output_dtype=..."` on anything else.                                                                                                                                           
  - New `SingleStreamDecoder::maybeConvertToFloat32` helper, called on every public frame-returning path: `getNextFrame`, `getFrameAtIndex`, `getFramesAtIndices`, `getFramesInRange`,     
  `getFramePlayedAt`, and both branches of `getFramesPlayedInRange`. Returns the tensor unchanged for `UINT8`; for `FLOAT32` it casts to float32 and normalizes by `maxVal`.
  - `torchcodec._core.ops.add_video_stream` gains an `output_dtype: str | None = None` kwarg that forwards to the custom op. That's the only Python surface that changes; `VideoDecoder`,
  `_frame.py`, and `test_decoders.py` are untouched on this branch.     

Follow-ups:
- CPU HDR output — `SwScale` / `FilterGraph` emit RGB48 for >8-bit sources, tensor allocation handles uint16
- Beta CUDA HDR output — NVDEC emits P016, NPP 16-bit color-twist converts to uint16 RGB, and `NVDECCache` keys on surface format.
- Modify API

## Test plan

- New unit tests in `test/test_ops.py` until we modify our API:
  - `test_output_dtype_float32_sdr` — float32 on an 8-bit source returns `float32` in `[0, 1]` and matches `uint8_ref / 255.0`.
  - `test_output_dtype_auto_sdr_is_uint8` — `"auto"` on SDR keeps `uint8` (no behavior change until step 2).
  - `test_output_dtype_invalid` — rejects unsupported dtypes (`"not_a_dtype"`, `blah`).
- Existing decoder tests pass on cpu / cuda / cuda:beta × exact / approximate seek modes.
- No change to default behavior: `VideoDecoder(path)[0].dtype == torch.uint8` remains true.

### CODE REVIEW

File: src/torchcodec/decoders/_video_decoder.py [hunk 1]
```diff
@@ -123,6 +123,11 @@ class VideoDecoder:
             :class:`~torchcodec.transforms.DecoderTransform` and
             :class:`~torchvision.transforms.v2.Transform`
             objects. Read more about this parameter in: TODO_DECODER_TRANSFORMS_TUTORIAL.
+        output_dtype (torch.dtype or str, optional): The dtype of decoded frame
+            tensors. One of ``torch.uint8`` (default), ``torch.float32``, or
+            ``"auto"``. ``float32`` returns tensors normalized to ``[0, 1]``.
+            ``auto`` returns ``uint8`` for SDR sources; support for higher bit
+            depths will be added in a follow-up.
```
Comment (Dan-Flores) [inline src/torchcodec/decoders/_video_decoder.py]: We probably don't need to promise more bit depths here, its not relevant to a user trying to understand our existing API.

File: src/torchcodec/decoders/_video_decoder.py [hunk 2]
```diff
@@ -271,11 +292,19 @@ def cpu_fallback(self) -> CpuFallbackStatus:
 
         return self._cpu_fallback
 
+    def _maybe_to_float32(self, tensor: Tensor) -> Tensor:
+        if self._output_dtype == torch.float32 or (
+            self._output_dtype == "auto" and tensor.dtype == torch.uint16
+        ):
+            max_val = 65535.0 if tensor.dtype == torch.uint16 else 255.0
```
Comment (Dan-Flores) [inline src/torchcodec/decoders/_video_decoder.py]: We can programatically get a dtype's max value using [torch.iinfo](https://docs.pytorch.org/docs/stable/type_info.html), which might be cleaner whan expanding this ternary if we ever support more dtypes:
```
torch.iinfo(tensor.dtype).max
```

File: test/test_decoders.py [hunk 3]
```diff
@@ -184,6 +184,29 @@ def test_create_fails(self):
...
 
+    def test_output_dtype_float32_sdr(self):
+        # float32 on an 8-bit source should be exactly (uint8_output / 255.0).
+
+        uint8_frame = VideoDecoder(NASA_VIDEO.path)[0]
+        float_frame = VideoDecoder(NASA_VIDEO.path, output_dtype=torch.float32)[0]
+
+        assert uint8_frame.dtype == torch.uint8
+        assert float_frame.dtype == torch.float32
+        torch.testing.assert_close(float_frame, uint8_frame.to(torch.float32) / 255.0)
```
Comment (Dan-Flores) [inline test/test_decoders.py]: Nice test!

File: src/torchcodec/decoders/_video_decoder.py [hunk 4]
```diff
@@ -271,11 +292,19 @@ def cpu_fallback(self) -> CpuFallbackStatus:
...
+            return tensor.to(torch.float32).div_(max_val)
+        return tensor
+
     def _getitem_int(self, key: int) -> Tensor:
         assert isinstance(key, int)
 
         frame_data, *_ = core.get_frame_at_index(self._decoder, frame_index=key)
-        return frame_data
+        return self._maybe_to_float32(frame_data)
```
Comment (Dan-Flores) [inline src/torchcodec/decoders/_video_decoder.py]: The Python API should be a very thin layer over the C++ backend code that implements the complex logic to make our code easier to reason about. Lets try to avoid adding `_maybe_to_float32()` here, since it requires us to update many callsites. 
Could `_maybe_to_float32()` live in `SingleStreamDecoder`, similar to `maybePermuteHWC2CHW`?

File: src/torchcodec/_core/StreamOptions.h [hunk 5]
```diff
@@ -21,6 +21,27 @@ enum ColorConversionLibrary {
...
+  switch (outputDtype) {
+    case OutputDtype::UINT8:
+      return 8;
+    case OutputDtype::FLOAT32:
+      return sourceBitDepth;
+    case OutputDtype::AUTO:
+      return (sourceBitDepth <= 8) ? 8 : sourceBitDepth;
+  }
+  return 8;
+}
```
Comment (Dan-Flores) [inline src/torchcodec/_core/StreamOptions.h]: This isn't used until https://github.com/meta-pytorch/torchcodec/pull/1343, lets remove it from this PR.

File: src/torchcodec/decoders/_video_decoder.py (hunk 4 above)
Comment (mollyxu) [reply src/torchcodec/decoders/_video_decoder.py]: Thanks for the suggestion! I think it is cleaner to have `_maybe_to_float32` follow the pattern of `maybePermuteHWC2CHW`

File: src/torchcodec/decoders/_video_decoder.py [hunk 6]
```diff
@@ -382,13 +400,17 @@ def get_frames_in_range(self, start: int, stop: int, step: int = 1) -> FrameBatc
...
             start=start,
             stop=stop,
             step=step,
         )
-        return FrameBatch(*frames)
+        return FrameBatch(
+            data=data,
+            pts_seconds=pts_seconds,
+            duration_seconds=duration_seconds,
+        )
```
Comment (Dan-Flores) [inline src/torchcodec/decoders/_video_decoder.py]: nit: lets reduce the diff here, I think this change was previously needed but not anymore. Same below on the other returned `FrameBatch`.

File: src/torchcodec/decoders/_video_decoder.py [hunk 7]
```diff
@@ -169,6 +173,7 @@ def __init__(
         custom_frame_mappings: (
             str | bytes | io.RawIOBase | io.BufferedReader | None
         ) = None,
+        output_dtype: "torch.dtype | str | None" = None,
```
Comment (Dan-Flores) [inline src/torchcodec/decoders/_video_decoder.py]: Since this PR only adds the plumbing, lets not expose this option in the public API yet. Adding the option to the public API should come last, after we've fully enabled the feature and tested that it works.

Lets also remove the updated docstrings, since we do not want this to be user facing just yet.

File: src/torchcodec/_frame.py [hunk 8]
```diff
@@ -70,7 +70,7 @@ class FrameBatch(Iterable):
     """
 
     data: Tensor
-    """The frames data (``torch.Tensor`` of uint8)."""
+    """The frames data (``torch.Tensor`` of uint8 by default; float32 in ``[0, 1]`` if ``output_dtype`` is ``torch.float32`` or ``"auto"``)."""
```
Comment (Dan-Flores) [inline src/torchcodec/_frame.py]: Lets hold off on these docstring changes as well until this feature is nearly complete.

**Review by Dan-Flores — COMMENTED** [review]
Nice work on this! Let's hold off on exposing this in the public API until this feature is complete. I think you will need to move the tests to `test_ops.py` once we remove the changes to `VideoDecoder`.

File: src/torchcodec/_core/_decoder_utils.py [hunk 9]
```diff
@@ -166,6 +166,7 @@ def create_video_decoder(
     device_variant: str = "ffmpeg",
     transforms: Sequence[DecoderTransform | nn.Module] | None = None,
     custom_frame_mappings: tuple[Tensor, Tensor, Tensor] | None = None,
+    output_dtype: str | None = None,
```
Comment (Dan-Flores) [inline src/torchcodec/_core/_decoder_utils.py]: Can we clean this up and add it as part of the public API updates PR, or is it needed now?

File: src/torchcodec/_core/SingleStreamDecoder.cpp [hunk 10]
```diff
@@ -1476,6 +1485,27 @@ torch::stable::Tensor SingleStreamDecoder::maybePermuteHWC2CHW(
   }
 }
 
+torch::stable::Tensor SingleStreamDecoder::maybeConvertToFloat32(
+    torch::stable::Tensor& tensor) {
+  OutputDtype outputDtype =
+      streamInfos_[activeStreamIndex_].videoStreamOptions.outputDtype;
+  bool isUInt16 = tensor.scalar_type() == torch::headeronly::ScalarType::UInt16;
+  bool shouldConvert = (outputDtype == OutputDtype::FLOAT32) ||
+      (outputDtype == OutputDtype::AUTO && isUInt16);
```
Comment (Dan-Flores) [inline src/torchcodec/_core/SingleStreamDecoder.cpp:1494]: I wonder if we could resolve `OutputDtype::AUTO` to an actual `OutputDtype` earlier, since `auto` just exists for user convenience. This might be easier once we get the `bitDepth` from frames in future PRs?

**Review by Dan-Flores — APPROVED** [review]

File: src/torchcodec/_core/SingleStreamDecoder.cpp (hunk 10 above)
Comment (mollyxu) [reply src/torchcodec/_core/SingleStreamDecoder.cpp:1494]: Yeah I think we can resolve `OutputDtype` earlier and I can do it in the next PR when we have `bitDepth` from frames
